### PR TITLE
Bugfix RestartFlow issue; #5165

### PR DESCRIPTION
**Pull request by Danthar**
For this PR i looked at what they did in the original bugfix on the JVM repo:
https://github.com/akka/akka/pull/24795/files#diff-ef6a29e6541e6af3e4452f69c790020e85116bad800388ce464dea7383eec9c0

And in whats currently in Master:
https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/RestartFlow.scala

Im not completely done yet. Still need to go through the Source and Sink implementations. And i want to see if i can add a UnitTest which verifies the behavior we want to fix.
But i wanted to get early feedback, if any.

I did ran into some things. It seems our Akka.Streams port could use some love. Lots of little changes that haven't been ported over the last year.
For example see https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/RestartFlow.scala#L425
The onDownstreamFinish has gotten a new Exception cause parameter. Which is absent in our codebase.

### CODE REVIEW

**Comment by Danthar** [comment]
I looked at why they added a delay in cancellation. It seems there is a missing feature which up to this day they havent realised yet, namely the propagation of a cancellation cause or something. Which would enable the RestartFlow to prevent it from conditionally propagating the cancellation or something. So i went with the same Delay stage that they up to this day still have implemented.
If its good enough for them, its probably good enough for us. 

Please check my implementation, my scala-fu could be better, so i might have missed something subtle that im not aware of.

File: src/core/Akka.Streams/Dsl/RestartFlow.cs
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Dsl/RestartFlow.cs:214]: we should file a separate issue for this once it's merged, so we can revisit in the future.

File: src/core/Akka.Streams/Dsl/RestartFlow.cs
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Dsl/RestartFlow.cs:482]: Agree - we should file an issue

File: src/core/Akka.Streams/Dsl/RestartSink.cs
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Dsl/RestartSink.cs:15]: When you say "completes" here - you don't mean that when the graph completes right? That wasn't clear from reading this.

**Review by Aaronontheweb — COMMENTED** [review]
Left some comments - need to make sure you do API approval though: https://getakka.net/community/public-api-changes.html

**Comment by Aaronontheweb** [comment]
Looks like the unit tests need to be updated too - lots of NREs

File: src/core/Akka.Streams/Dsl/RestartFlow.cs
Comment (Danthar) [reply src/core/Akka.Streams/Dsl/RestartFlow.cs:202]: Thats one thing im unsure about. 
In the scala source, the `Delay` class is a case class, wrapped in an object. (see: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/RestartFlow.scala#L402)  Im unsure if that means its a public api or not. But i figured undefined access modifiers means private/internal so i went with that.

The comments on the original scala bugfix suggest its overridable through a WithAttributes call on the RestartFlow, but im not sure if that works here (regardless of the access modifier). im pretty sure they are not applied to the flow the RestartFlow is wrapping.

However i think normally nobody will ever want to tweak stuff in their stream on this level. Im of the opinion that "nobody should ever need to tweak this, because it should just work". But considering what the fix is, they might have to ? Although i'd expect that the # of ppl that would need to do this would be vanishingly rare.

File: src/core/Akka.Streams/Dsl/RestartFlow.cs
Comment (Danthar) [reply src/core/Akka.Streams/Dsl/RestartFlow.cs:214]: Agreed. Ill make a note for myself

File: src/core/Akka.Streams/Dsl/RestartSink.cs
Comment (Danthar) [reply src/core/Akka.Streams/Dsl/RestartSink.cs:15]: No idea. I only seperated the Sink and Source code to seperate files at this point. Still need to go through them. Ill take a closer look at the comments then as well.

**Comment by Danthar** [comment]
Thx for the review. Answered some of your remarks. Yup the Units Tests needs work, ignored those for now. My next step is first going through the Source and Sink codebase to port their latest changes. And then im going to look at the UnitTests.
I expect that fixing those will take most of the time for this PR.

File: src/core/Akka.Streams/Dsl/RestartFlow.cs
Comment (Aaronontheweb) [reply src/core/Akka.Streams/Dsl/RestartFlow.cs:202]: Yeah that's a public API - it's meant to be an attribute you can pass in at the time the flow is declared via the `inheritedAttributes` collection. So I guess, broadly speaking, this is meant to be configurable via the attributes approach.

Comment (Aaronontheweb) [reply src/core/Akka.Streams/Dsl/RestartFlow.cs:202]: > Although i'd expect that the # of ppl that would need to do this would be vanishingly rare.

Agree - so the duration here just sets a 50ms threshold for restarting. That seems fine for most cases, where it's recovering from a crash.

File: src/core/Akka.Streams/Dsl/RestartSource.cs
Comment (Danthar) [inline src/core/Akka.Streams/Dsl/RestartSource.cs:176]: Brought this in line with the scala sourcecode. IMO this used to contain a big ass bug. Before it never used the SourceFactory. How that ever worked i don't know. I can only guess the specs surrounding the RestartSource are bad.

File: src/core/Akka.Streams/Dsl/RestartSink.cs
Comment (Danthar) [reply src/core/Akka.Streams/Dsl/RestartSink.cs:15]: The docs are the same as on the scala side. I think what is being referenced here is the stage itself. And not the entire graph.

**Comment by Danthar** [comment]
Ok, im looking at this bit of code in the Tests.
https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs#L691

While im sure its very handy to verify that a flow wrapped by the RestartFlow is actually restarted. I think it does nothing to verify that a message supplied to the graph of which the RestartFlow stage is a part, is actually Retried without having the entire graph being completed/failed.

I could not find the original unit tests in the scala repo btw, i wonder how they test this.

**Comment by Aaronontheweb** [comment]
> Ok, im looking at this bit of code in the Tests.
> https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs#L691
> 
> While im sure its very handy to verify that a flow wrapped by the RestartFlow is actually restarted. I think it does nothing to verify that a message supplied to the graph of which the RestartFlow stage is a part, is actually Retried without having the entire graph being completed/failed.
> 
> I could not find the original unit tests in the scala repo btw, i wonder how they test this.

I guess the easiest way to do it would be to pass a numbered sequence 1-10, set number 5 to cause a crash inside the original flow, and see if [1,4] [6,10] still arrive before the stream completes.

File: src/core/Akka.Streams/Dsl/RestartSource.cs
Comment (Aaronontheweb) [reply src/core/Akka.Streams/Dsl/RestartSource.cs:176]: Given how racy those specs have been, I think that's a good assumption

**Comment by Aaronontheweb** [comment]
@Danthar looks like the only specs that are failing right now are the API Approval specs https://getakka.net/community/public-api-changes.html

**Comment by Danthar** [comment]
> 
> I guess the easiest way to do it would be to pass a numbered sequence 1-10, set number 5 to cause a crash inside the original flow, and see if [1,4] [6,10] still arrive before the stream completes.

Not quite. Since it should keep retrying the original message. So the error on number 5 should be transient, as in, only occur a few times. And then the flow should keep retrying untill it succeeds and end up emitting the entire sequence at the end.

I still want to add a test which verifies this. But that will be next week probably before i can get back to this.

**Comment by Danthar** [comment]
Bad news. Playing around with a unit test shows it does not work. Now somehow it seems like the stage is not restarting.

File: src/core/Akka.Streams/Dsl/RestartFlow.cs
Comment (Danthar) [reply src/core/Akka.Streams/Dsl/RestartFlow.cs:485]: Done

File: src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs
Comment (Aaronontheweb) [inline src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs:1078]: Nice catch

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**
